packages/venia-concept/src/ServiceWorker/Utilities/htmlHandler.js

```diff
@@ -1,7 +1,144 @@
 import { HTML_UPDATE_AVAILABLE } from '@magento/venia-ui/lib/constants/swMessageTypes';
+import { parse } from 'node-html-parser';
 
 import { sendMessageToWindow } from './messageHandler';
 
+const generateScriptResource = htmlElement => htmlElement.attributes.src;
+
+const generateLinkResource = htmlElement => htmlElement.attributes.href;
+
+const generateStyleResource = htmlElement => htmlElement.text;
+
+const identity = x => x;
+
+/**
+ * Generates list of resources from the parsed HTML object.
+ *
+ * @param {HTMLElement} parsedHTML
+ *
+ * @returns { scripts: [string], links: [string], styles: [string] }
+ */
+const generateResources = parsedHTML => {
+    const resources = {};
+
+    resources.scripts = parsedHTML
+        .querySelectorAll('script')
+        .map(generateScriptResource)
+        .filter(identity)
+        .sort();
+
+    resources.links = parsedHTML
+        .querySelectorAll('link')
+        .map(generateLinkResource)
+        .filter(identity)
+        .sort();
+
+    resources.styles = parsedHTML
+        .querySelectorAll('style')
+        .map(generateStyleResource)
+        .filter(identity)
+        .sort();
+
+    return resources;
+};
+
+/**
+ * Returns true if resouce url/urls have changed.
+ *
+ * @param {HTMLElement} newDOM
+ * @param {HTMLElement} oldDOM
+ *
+ * @returns Boolean
+ */
+const haveResourcesChanged = (newDOM, oldDOM) => {
+    const newResources = generateResources(newDOM);
+    const oldResources = generateResources(oldDOM);
+    const keysToCompare = Object.keys(newResources);
+
+    return !keysToCompare.reduce((acc, key) => {
+        const hasResourceChanged =
+            newResources[key].toString() === oldResources[key].toString();
+
+        return acc && hasResourceChanged;
+    }, true);
+};
+
+/**
+ * Returns true if title has changed.
+ *
+ * @param {HTMLElement} newDOM
+ * @param {HTMLElement} oldDOM
+ *
+ * @returns Boolean
+ */
+const hasTitleChanged = (newDOM, oldDOM) => {
+    const oldTitle = (oldDOM.querySelector('title') || {}).text;
+    const newTitle = (newDOM.querySelector('title') || {}).text;
+
+    return newTitle !== oldTitle;
+};
+
+/**
+ * Returns true if media backend url has changed.
+ *
+ * @param {HTMLElement} newDOM
+ * @param {HTMLElement} oldDOM
+ *
+ * @returns Boolean
+ */
+const hasBackendUrlChanged = (newDOM, oldDOM) => {
+    const oldUrl = oldDOM.querySelector('html').attributes[
+        'data-media-backend'
+    ];
+    const newUrl = newDOM.querySelector('html').attributes[
+        'data-media-backend'
+    ];
```

Comment:
There are selectors for data attributes. See here: https://css-tricks.com/almanac/selectors/a/attribute/.

This could be `querySelector('[data-media-backend]')` if you want.

Reply:
We can not use such selectors because the type of the dom object, in this case, is actually not what `DOMParser` returns, this object is different and is returned by the `node-html-parser` which does not support complex selectors.

Reply:
We can not use such selectors because the type of the dom object, in this case, is actually not what DOMParser returns, this object is different and is returned by the node-html-parser which does not support complex selectors.